scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/license/LicenseSchedulingPolicy.java

```diff
@@ -168,39 +156,41 @@ private boolean acquireTokensForJob(JobDescriptorImpl job) {
 
         // Retrieve required licenses names from the job generic informations
         final String requiredLicenses = job.getInternal().getRuntimeGenericInformation().get(REQUIRED_LICENSES);
+        final String jobId = job.getJobId().value();
 
         // If it requires software licenses
         if (requiredLicenses != null) {
 
-            logger.debug("Considering licenses " + requiredLicenses);
+            logger.debug("For job considering licenses " + requiredLicenses);
 
             // Do not execute if one of the required license can not be obtained
             final HashSet<String> requiredLicensesSet = new HashSet<String>(Arrays.asList(requiredLicenses.split(",")));
             for (String requiredLicense : requiredLicensesSet) {
 
-                if (!remainingTokens.containsKey(requiredLicense)) {
+                if (!licenseSynchronization.containsLicense(requiredLicense)) {
                     logger.debug("The required software is not specified in the license properties file, keep job pending");
                     return false;
                 }
 
-                if (!containsJob(eligibleJobsDescriptorsLicenses.get(requiredLicense), job.getJobId()) &&
-                    !canGetToken(requiredLicense)) {
+                if (!licenseSynchronization.containsJobId(requiredLicense, jobId) && !canGetToken(requiredLicense)) {
                     logger.debug("Token for " + requiredLicense + " not available, keep job pending");
                     return false;
                 }
             }
             // Can get a new token for each required license !
             for (String requiredLicense : requiredLicensesSet) {
-                if (!containsJob(eligibleJobsDescriptorsLicenses.get(requiredLicense), job.getJobId())) {
-                    eligibleJobsDescriptorsLicenses.get(requiredLicense).add(job);
-                    remainingTokens.put(requiredLicense, remainingTokens.get(requiredLicense) - 1);
+                if (!licenseSynchronization.containsJobId(requiredLicense, jobId)) {
+                    licenseSynchronization.addJobToLicense(requiredLicense, jobId);
+                    licenseSynchronization.acquireToken(requiredLicense);
```

Comment:
it looks like addJobToLicense and acquireToken could be done in a single call (the second being called from the first)

Reply:
Ok.